File: src/main/java/fr/cph/crypto/controller/UserController.java
package fr.cph.crypto.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

import fr.cph.crypto.domain.Position;
import fr.cph.crypto.domain.User;
import fr.cph.crypto.repository.UserRepository;
import fr.cph.crypto.service.UserService;

@RequestMapping(value = "/user")
@RestController
public class UserController {

	private final UserRepository repository;
	private final UserService userService;

	@Autowired
	public UserController(final UserRepository repository, UserService userService) {
		this.repository = repository;
		this.userService = userService;
	}

	@RequestMapping
	public List<User> getAll() {
		return StreamSupport.stream(repository.findAll().spliterator(), false).collect(Collectors.toList());
	}

	@RequestMapping(value = "/{id}", method = RequestMethod.GET)
	public User getUser(@PathVariable("id") final String id) {
		return repository.findOne(id);
	}

	@RequestMapping(value = "/{id}/position/{positionId}", method = RequestMethod.PUT)
	public Position updatePosition(@PathVariable("id") final String id, @RequestBody final Position position) {
		return userService.updatePosition(position);
	}
}

File: src/main/kotlin/fr/cph/crypto/controller/UserController.kt
package fr.cph.crypto.controller

import fr.cph.crypto.domain.Position
import fr.cph.crypto.domain.User
import fr.cph.crypto.repository.UserRepository
import fr.cph.crypto.service.UserService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import java.security.Principal

@RequestMapping(value = ["/api/user"])
@RestController
class UserController @Autowired
constructor(private val repository: UserRepository, private val userService: UserService) {

    @PreAuthorize("hasAuthority('ADMIN')")
    @RequestMapping(method = [RequestMethod.GET])
    fun getAllUsers(@RequestHeader(value = "Authorization") bearerToken: String): List<User> {
        return repository.findAll().toList()
    }

    @RequestMapping(method = [RequestMethod.POST])
    fun createUser(@RequestBody user: User): User {
        return userService.createUser(user)
    }

    @RequestMapping(value = ["/{id}"], method = [RequestMethod.GET])
    fun getUser(@PathVariable("id") id: String, principal: Principal): User {
        val user = repository.findOne(id)
        if (principal.name == user.email) {
            return user
        } else {
            throw RuntimeException()
        }
    }

    @RequestMapping(value = ["/{id}/position/refresh"], method = [RequestMethod.GET])
    fun refreshUser(@PathVariable("id") id: String): List<Position> {
        return userService.refreshUserPositions(id)
    }

    @RequestMapping(value = ["/{id}/position/{positionId}"], method = [RequestMethod.PUT])
    fun updatePosition(@PathVariable("id") id: String, @RequestBody position: Position): Position {
        return userService.updatePosition(position)
    }
}
